File: C++/015. 3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) 
    {
        int n = nums.size();
        if ( n < 3) return {};
        vector<vector<int>> res = {};
        sort(nums.begin(), nums.end());
        int i = 0;
        while(i < n-2)
        {
            int left = i + 1;
            int right = n - 1;
            while(left < right)
            {
                if (nums[left] + nums[right] < -1*nums[i])
                {
                    left++;
                }
                else if (nums[left] + nums[right] > -1*nums[i])
                {
                    right--;
                }
                else
                {
                    res.push_back({nums[i], nums[left], nums[right]});
                    int a = nums[left];
                    int b = nums[right];
                    while(left < right && nums[left] == a){left++;}
                    while(right > left && nums[right] == b){right--;}
                }
            }
            int c = nums[i];
            while(i < n-2 && nums[i] == c){i++;}
        }
        return res;
    }
};

File: python/015. 3Sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = len(nums)
        if n < 3: return []
        nums.sort()
        res = []
        i = 0
        while i <= n-3:
            left = i+1
            right = n-1
            num = nums[i]
            while left < right:
                if nums[left] + nums[right] < - num: left += 1
                elif nums[left] + nums[right] > - num: right -= 1
                else: 
                    res.append([num, nums[left], nums[right]])
                    a, b = nums[left], nums[right]
                    while left < right and nums[left] == a: left += 1
                    while left < right and nums[right] == b: right -= 1
            while i <= n-3 and nums[i] == num: i += 1
        return res

